ttnn/cpp/ttnn/operations/normalization/softmax/softmax_compiler_interface.hpp

```diff
@@ -0,0 +1,63 @@
+// SPDX-FileCopyrightText: © 2023 Tenstorrent Inc.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#pragma once
+
+#include <exception>
+#include <functional>
+
+#include "third_party/json/json.hpp"
+#include "tt_metal/common/logger.hpp"
+#include "ttnn/compiler_interface/compiler_interface.hpp"
+#include "ttnn/graph/graph_processor.hpp"
+#include "ttnn/graph/graph_trace_utils.hpp"
+#include "ttnn/operations/normalization/softmax/softmax.hpp"
+
+namespace ttnn::compiler_interface {
+
+std::tuple<bool, size_t, size_t, size_t> softmax_op_constraints(
+    const OperandParams& input, const int dim_arg, const OperandParams& output) {
+    // get_op_trace is a lambda that prepares input and output tensors, capturing graph trace of the op without
+    // inputs allocation.
+    auto get_op_trace = [](const OperandParams& input, const int dim_arg, const OperandParams& output) {
+        nlohmann::json op_trace;
+
+        // outer graph capture is used to avoid capturing and dispatching of dummy input tensor(s) creation
+        {
+            ttnn::graph::GraphProcessor::begin_graph_capture(ttnn::graph::GraphProcessor::RunMode::NO_DISPATCH);
+            const auto input_tensor = create_device_tensor(
+                std::get<ttnn::SimpleShape>(input),
+                std::get<tt::tt_metal::DataType>(input),
+                std::get<tt::tt_metal::Layout>(input),
+                SingletonDeviceContext::get_instance().get_device(),
```

Comment:
You shared about this before, but I am still not clear why is this needed. 
Can you please share more? We can hop on a call to discuss this.

- I assume its like a mock device so that real device is not needed, or is it something else?
- Whats the point of singleton?

Reply:
Yes a mockup device allows us to run without  a real device. The Runtime team is still working on it, this branch contains my hacky implementation. 

This singleton is to avoid opening and closing mockup device during the compile time. In my implementation that still takes 200-300ms. We'll - see about the final implementation from runtime team. We aim to be under 1ms per query. 

To keep this interface clean, I can move this to the compiler and add a device pointer as an argument?